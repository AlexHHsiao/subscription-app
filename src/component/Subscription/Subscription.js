import React, {createRef, useEffect, useCallback, useState} from 'react';

const planSelection = createRef();
const seatInput = createRef();
const priceInput = createRef();

const Subscription = ({subscriptionData, getPreview, getUpdate, previewPrice}) => {
    const [updateBtnDisable, setUpdateBtnDisable] = useState(true);
    useEffect(() => {
        if (previewPrice) {
            priceInput.current.value = previewPrice;
        } else {
            planSelection.current.value = subscriptionData.plan;
            seatInput.current.value = subscriptionData.seats;
            priceInput.current.value = subscriptionData.cost;
            setUpdateBtnDisable(true);
        }
    }, [subscriptionData, previewPrice, setUpdateBtnDisable]);

    const onSubscriptionChange = useCallback(() => {
        if (isNaN(parseInt(seatInput.current.value, 10))) {
            seatInput.current.value = 1;
        }

        if (planSelection.current.value !== subscriptionData.plan || seatInput.current.value != subscriptionData.seats) {
            getPreview(planSelection.current.value, seatInput.current.value);
            setUpdateBtnDisable(false);
        } else {
            priceInput.current.value = subscriptionData.cost;
            setUpdateBtnDisable(true);
        }
    }, [subscriptionData, getPreview]);

    const onUpdate = useCallback(() => {
        getUpdate(planSelection.current.value, seatInput.current.value);
    }, [getUpdate]);

    return (
        <div className='card'>
            <div className='card-header'>
                Subscription
            </div>
            <div className='card-body'>
                <div className='row'>
                    <div className='col-sm-12 col-md-5'>
                        <div className='form-group'>
                            <label>Plan</label>
                            <select className='form-control' ref={planSelection}
                                    onChange={onSubscriptionChange}>
                                <option value='basic'>Basic</option>
                                <option value='good'>Good</option>
                                <option value='better'>Better</option>
                                <option value='best'>Best</option>
                            </select>
                        </div>
                    </div>
                    <div className='col-sm-12 col-md-5'>
                        <div className='form-group'>
                            <label>Seats</label>
                            <input type='number' className='form-control' ref={seatInput}
                                   onChange={onSubscriptionChange} min={1}
                            />
                        </div>
                    </div>
                    <div className='col-sm-12 col-md-2'>
                        <div className='form-group'>
                            <div className='form-group'>
                                <label>Price</label>
                                <input type='number' className='form-control' disabled
                                       ref={priceInput}
                                />
                            </div>
                        </div>
                    </div>
                </div>

                <button className='btn btn-primary float-right' onClick={onUpdate}
                        disabled={updateBtnDisable}
                >
                    Update Subscription
                </button>
            </div>
        </div>
    );
};

export default Subscription;
