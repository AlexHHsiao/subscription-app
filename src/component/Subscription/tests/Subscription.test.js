import React from 'react';
import {render} from '@testing-library/react';
import Subscription from '../Subscription';

const mockData = {
    subscriptionData: {
        plan: 'best',
        name: 'Best',
        seats: 1,
        cost: 50
    },
    getPreview: () => {},
    getUpdate: () => {},
    previewPrice: () => {}
};

test('<Subscription/>', () => {
    const {container: {firstChild}} = render(<Subscription {...mockData}/>);
    expect(firstChild).toMatchSnapshot();
});
