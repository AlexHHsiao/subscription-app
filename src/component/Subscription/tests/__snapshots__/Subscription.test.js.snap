// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Subscription/> 1`] = `
<div
  class="card"
>
  <div
    class="card-header"
  >
    Subscription
  </div>
  <div
    class="card-body"
  >
    <div
      class="row"
    >
      <div
        class="col-sm-12 col-md-5"
      >
        <div
          class="form-group"
        >
          <label>
            Plan
          </label>
          <select
            class="form-control"
          >
            <option
              value="basic"
            >
              Basic
            </option>
            <option
              value="good"
            >
              Good
            </option>
            <option
              value="better"
            >
              Better
            </option>
            <option
              value="best"
            >
              Best
            </option>
          </select>
        </div>
      </div>
      <div
        class="col-sm-12 col-md-5"
      >
        <div
          class="form-group"
        >
          <label>
            Seats
          </label>
          <input
            class="form-control"
            min="1"
            type="number"
          />
        </div>
      </div>
      <div
        class="col-sm-12 col-md-2"
      >
        <div
          class="form-group"
        >
          <div
            class="form-group"
          >
            <label>
              Price
            </label>
            <input
              class="form-control"
              disabled=""
              type="number"
            />
          </div>
        </div>
      </div>
    </div>
    <button
      class="btn btn-primary float-right"
      disabled=""
    >
      Update Subscription
    </button>
  </div>
</div>
`;
